internal/promote/storage.go

```diff
@@ -0,0 +1,410 @@
+			return sort.StringsAreSorted([]string{prs[i].Name, prs[j].Name})
+		}
+		return prs[i].semver.LessThan(&prs[j].semver)
+	})
+	return prs
+}
+
+// Strings method returns an array of string representations.
+func (prs PackageRevisions) Strings() []string {
+	var entries []string
+	for _, pr := range prs {
+		entries = append(entries, pr.String())
+	}
+	return entries
+}
+
+// CloneRepository method clones the repository and changes branch to stage.
+func CloneRepository(stage string) (*git.Repository, error) {
+	r, err := git.Clone(memory.NewStorage(), memfs.New(), &git.CloneOptions{
+		URL:           fmt.Sprintf(repositoryURL, "elastic"),
+		RemoteName:    remoteName,
+		ReferenceName: plumbing.NewBranchReferenceName(stage),
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "cloning package-storage repository failed")
+	}
+
+	err = r.Fetch(&git.FetchOptions{
+		RemoteName: remoteName,
+		RefSpecs:   []config.RefSpec{"refs/*:refs/*", "HEAD:refs/heads/HEAD"},
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "fetch remote branches failed")
+	}
+
+	user, err := User(r)
+	if err != nil {
+		return nil, errors.Wrap(err, "reading Git user failed")
+	}
+
+	_, err = r.CreateRemote(&config.RemoteConfig{
+		Name: user,
+		URLs: []string{
+			fmt.Sprintf(repositoryURL, user),
+		},
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "creating remote failed")
+	}
+	return r, nil
+}
+
+// ListPackages method lists available packages in the package-storage.
+// It skips technical packages (snapshot, staging).
+func ListPackages(r *git.Repository) (PackageRevisions, error) {
+	wt, err := r.Worktree()
+	if err != nil {
+		return nil, errors.Wrap(err, "fetching worktree reference failed")
+	}
+
+	packageDirs, err := wt.Filesystem.ReadDir("/packages")
+	if err != nil {
+		return nil, errors.Wrap(err, "reading packages directory failed")
+	}
+
+	var revisions PackageRevisions
+	for _, packageDir := range packageDirs {
+		if !packageDir.IsDir() {
+			continue
+		}
+
+		if packageDir.Name() == snapshotPackage || packageDir.Name() == stagingPackage {
+			continue
+		}
+
+		versionDirs, err := wt.Filesystem.ReadDir(filepath.Join("/packages", packageDir.Name()))
+		if err != nil {
+			return nil, errors.Wrap(err, "reading packages directory failed")
+		}
+
+		for _, versionDir := range versionDirs {
+			if !versionDir.IsDir() {
+				continue
+			}
+
+			packageVersion, err := semver.NewVersion(versionDir.Name())
+			if err != nil {
+				return nil, errors.Wrapf(err, "reading package version failed (name: %s, version: %s)", packageDir.Name(), versionDir.Name())
+			}
+
+			revisions = append(revisions, PackageRevision{
+				Name:    packageDir.Name(),
+				Version: versionDir.Name(),
+				semver:  *packageVersion,
+			})
+		}
+	}
+	return revisions.sort(), nil
+}
+
+// FilterPackages method filters package revisions based on the "newest revision only" policy.
+func FilterPackages(allPackages PackageRevisions, newestOnly bool) PackageRevisions {
```

Comment:
This function looks like another candidate to become a method with a `PackageRevisions` receiver.

Reply:
Fixed